### PR TITLE
Configure Clam AV

### LINKED ISSUE #1237: Add ClamAV definition mirror
## :strawberry: Desired behaviour

ClamAV requests that you use your own mirror if multiple machines are trying to retrieve updates. They recommend [cvdupdate](https://pypi.org/project/cvdupdate/)

## :steam_locomotive: Possible workarounds, remediations or solutions

This is requested but not required

### LINKED ISSUE #1783: Add ClamAV to the workspaces
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a request for a new feature in the Data Safe Haven or an upgrade to an existing feature.
- [x] The feature is still missing in the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :strawberry: Suggested change

The workspace VMs currently have no antivirus, we should fix this.

## :steam_locomotive: How could this be done?

Install ClamAV and add appropriate NSG/firewall rules to allow it to work as intended.

### PR DESCRIPTION
**Pull request by JimMadge**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :vertical_traffic_light: Depends on

#1892 

### :arrow_heading_up: Summary

- Add Clam AV configuration, services and active services.
- Adds a ClamAV mirror using a Docker image built from [here](https://github.com/chmey/docker-clamav-mirror)
  - we may want to use our own Docker image in future

### :closed_umbrella: Related issues

Closes #1237
Closes #1783

### :microscope: Tests

Testing in progress

### CODE REVIEW

**Comment by JimMadge** [comment]
Freshclam is having a problem fetching updated definitions.

```command
Tue Jul 16 15:09:01 2024 -> daily database available for download (remote version: 27338)
Tue Jul 16 15:09:01 2024 -> ^Download failed (35) Tue Jul 16 15:09:01 2024 -> ^ Message: SSL connect error
Tue Jul 16 15:09:01 2024 -> ^Can't download daily.cvd from https://database.clamav.net/daily.cvd
Tue Jul 16 15:09:01 2024 -> Trying again in 5 secs...
```

Clamav made changes to their servers so that _only_ freshclam or cvdupdate can fetch updates, through checking headers. It is possible that the Azure firewall is dropping the required headers here.

**Comment by jemrobinson** [comment]
Can you test whether this is the case (e.g. by temporarily exempting the workspace subnet from routing via the firewall)?

**Comment by JimMadge** [comment]
Putting some references here for the future,

- https://docs.clamav.net/faq/faq-eol.html#version-support-matrix
- https://github.com/Cisco-Talos/clamav/issues/1287

**Comment by jemrobinson** [comment]
Worth using a mirror? https://docs.clamav.net/appendix/CvdPrivateMirror.html. Could have this as e.g. a Docker image served via an ACI.

**Comment by JimMadge** [comment]
That's a great idea. Similar to what we already do with apt/PyPI/CRAN

File: data_safe_haven/resources/workspace/workspace.cloud_init.mustache.yaml
Comment (JimMadge) [inline data_safe_haven/resources/workspace/workspace.cloud_init.mustache.yaml:7]: Would be good to do this in Ansible. However, as there are other instances in cloud init let's open a new issue for that.

File: data_safe_haven/resources/workspace/ansible/host_vars/localhost.yaml
Comment (JimMadge) [inline data_safe_haven/resources/workspace/ansible/host_vars/localhost.yaml]:
```suggestion
      - clamav-daemon
```

These are dependencies of `clamav` and `clamav-daemon`.

(I thought when I tested this `clamav` was providing `clamd` and `clamdscan` but [this](https://packages.ubuntu.com/jammy/utils/clamav) suggests that isn't true).

File: data_safe_haven/resources/workspace/ansible/desired_state.yaml
Comment (JimMadge) [inline data_safe_haven/resources/workspace/ansible/desired_state.yaml]:
```suggestion
        line: "{{ lookup('file', '/etc/clamav/freshclam-mirror.conf') }}"
```

Jinja2 likes spaces inside braces. I think it can fail sometimes if you don't.

This will read the contents of `freshclam-mirror.conf` and write it to `freshclam.conf`?

File: data_safe_haven/resources/workspace/ansible/desired_state.yaml
Comment (JimMadge) [inline data_safe_haven/resources/workspace/ansible/desired_state.yaml]:
```suggestion
        create: yes
        owner: root
        group: root
        mode: '0644'
```

If we have create we should make sure the user, group and permissions are correct.

File: data_safe_haven/types/enums.py
Comment (JimMadge) [inline data_safe_haven/types/enums.py:169]: What does HKP stand for?

**Review by JimMadge — COMMENTED** [review]
LGTM (especially my parts 🙃).

I can't approve or request changes as it is my PR. I've made some suggestions which I think should be resolved before merging.

File: data_safe_haven/resources/workspace/workspace.cloud_init.mustache.yaml
Comment (jemrobinson) [reply data_safe_haven/resources/workspace/workspace.cloud_init.mustache.yaml:7]: We need a way to get these variables into the desired state file. This could be uploading a file with variables into blob storage or it could be writing them to the cloud-init. As you say, this is done in a few other places so that's the pattern I copied.

Comment (jemrobinson) [reply data_safe_haven/resources/workspace/workspace.cloud_init.mustache.yaml:7]: Captured in #2075

File: data_safe_haven/resources/workspace/ansible/desired_state.yaml
Comment (jemrobinson) [reply data_safe_haven/resources/workspace/ansible/desired_state.yaml]: Yes, we want to append `PrivateMirror clamav.$SRE_FQDN` to this file, but SRE_FQDN isn't available to Ansible as a variable. Could be refactored as part of #2075.

File: data_safe_haven/types/enums.py
Comment (jemrobinson) [reply data_safe_haven/types/enums.py:169]: Http Keyserver Protocol: https://en.wikipedia.org/wiki/Key_server_(cryptographic)

**Comment by jemrobinson** [comment]
The `clamav-freshclam` service is not using the new file after the config file is changed. Logs are below, showing two attempts, the first at deployment time and the second an hour later. I assume the issue is in the notify/handler logic. Any thoughts @JimMadge ?

N.B. manually running `systemctl restart clamav-freshclam` triggers the update as expected.

```
$ cat /etc/clamav/freshclam.conf 
# Automatically created by the clamav-freshclam postinst
# Comments will get lost when you reconfigure the clamav-freshclam package

DatabaseOwner clamav
UpdateLogFile /var/log/clamav/freshclam.log
LogVerbose false
LogSyslog false
LogFacility LOG_LOCAL6
LogFileMaxSize 0
LogRotate true
LogTime true
Foreground false
Debug false
MaxAttempts 5
DatabaseDirectory /var/lib/clamav
DNSDatabaseInfo current.cvd.clamav.net
ConnectTimeout 30
ReceiveTimeout 0
TestDatabases yes
ScriptedUpdates yes
CompressLocalDatabase no
Bytecode true
NotifyClamd /etc/clamav/clamd.conf
# Check for new database 24 times a day
Checks 24
DatabaseMirror db.local.clamav.net
DatabaseMirror database.clamav.net
PrivateMirror clamav.fuschia.pink.develop.turingsafehaven.ac.uk
```

```
$ cat /var/log/clamav/freshclam.log 
cat /var/log/clamav/freshclam.log 
Fri Aug  2 17:42:40 2024 -> --------------------------------------
Fri Aug  2 17:42:41 2024 -> freshclam daemon 0.103.11 (OS: linux-gnu, ARCH: x86_64, CPU: x86_64)
Fri Aug  2 17:42:41 2024 -> ClamAV update process started at Fri Aug  2 17:42:41 2024
Fri Aug  2 17:42:41 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 17:42:41 2024 -> WARNING: Download failed (35) Fri Aug  2 17:42:41 2024 -> WARNING:  Message: SSL connect error
Fri Aug  2 17:42:41 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 17:42:41 2024 -> Trying again in 5 secs...
Fri Aug  2 17:42:46 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 17:42:46 2024 -> WARNING: Download failed (35) Fri Aug  2 17:42:46 2024 -> WARNING:  Message: SSL connect error
Fri Aug  2 17:42:46 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 17:42:46 2024 -> Trying again in 5 secs...
Fri Aug  2 17:42:51 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 17:42:51 2024 -> WARNING: Download failed (35) Fri Aug  2 17:42:51 2024 -> WARNING:  Message: SSL connect error
Fri Aug  2 17:42:51 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 17:42:51 2024 -> Trying again in 5 secs...
Fri Aug  2 17:42:56 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 17:42:56 2024 -> WARNING: Download failed (35) Fri Aug  2 17:42:56 2024 -> WARNING:  Message: SSL connect error
Fri Aug  2 17:42:56 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 17:42:56 2024 -> Trying again in 5 secs...
Fri Aug  2 17:43:01 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 17:43:01 2024 -> ERROR: Download failed (35) Fri Aug  2 17:43:01 2024 -> ERROR:  Message: SSL connect error
Fri Aug  2 17:43:01 2024 -> ERROR: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 17:43:01 2024 -> Giving up on https://database.clamav.net...
Fri Aug  2 17:43:01 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 17:43:01 2024 -> WARNING: Download failed (35) Fri Aug  2 17:43:01 2024 -> WARNING:  Message: SSL connect error
Fri Aug  2 17:43:01 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 17:43:01 2024 -> Trying again in 5 secs...
Fri Aug  2 17:43:06 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 17:43:06 2024 -> WARNING: Download failed (35) Fri Aug  2 17:43:06 2024 -> WARNING:  Message: SSL connect error
Fri Aug  2 17:43:06 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 17:43:06 2024 -> Trying again in 5 secs...
Fri Aug  2 17:43:11 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 17:43:11 2024 -> WARNING: Download failed (35) Fri Aug  2 17:43:11 2024 -> WARNING:  Message: SSL connect error
Fri Aug  2 17:43:11 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 17:43:11 2024 -> Trying again in 5 secs...
Fri Aug  2 17:43:16 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 17:43:16 2024 -> WARNING: Download failed (35) Fri Aug  2 17:43:16 2024 -> WARNING:  Message: SSL connect error
Fri Aug  2 17:43:16 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 17:43:16 2024 -> Trying again in 5 secs...
Fri Aug  2 17:43:21 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 17:43:21 2024 -> ERROR: Download failed (35) Fri Aug  2 17:43:21 2024 -> ERROR:  Message: SSL connect error
Fri Aug  2 17:43:21 2024 -> ERROR: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 17:43:21 2024 -> Giving up on https://database.clamav.net...
Fri Aug  2 17:43:21 2024 -> ERROR: Update failed for database: daily
Fri Aug  2 17:43:21 2024 -> ERROR: Database update process failed: Connection failed
Fri Aug  2 17:43:21 2024 -> ERROR: Update failed.
Fri Aug  2 17:43:21 2024 -> --------------------------------------
Fri Aug  2 18:43:21 2024 -> Received signal: wake up
Fri Aug  2 18:43:21 2024 -> ClamAV update process started at Fri Aug  2 18:43:21 2024
Fri Aug  2 18:43:21 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 18:43:21 2024 -> WARNING: Download failed (35) Fri Aug  2 18:43:21 2024 -> WARNING:  Message: SSL connect error
Fri Aug  2 18:43:21 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 18:43:21 2024 -> Trying again in 5 secs...
Fri Aug  2 18:43:26 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 18:43:26 2024 -> WARNING: Download failed (35) Fri Aug  2 18:43:26 2024 -> WARNING:  Message: SSL connect error
Fri Aug  2 18:43:26 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 18:43:26 2024 -> Trying again in 5 secs...
Fri Aug  2 18:43:31 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 18:43:31 2024 -> WARNING: Download failed (35) Fri Aug  2 18:43:31 2024 -> WARNING:  Message: SSL connect error
Fri Aug  2 18:43:31 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 18:43:31 2024 -> Trying again in 5 secs...
Fri Aug  2 18:43:36 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 18:43:36 2024 -> WARNING: Download failed (35) Fri Aug  2 18:43:36 2024 -> WARNING:  Message: SSL connect error
Fri Aug  2 18:43:36 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 18:43:36 2024 -> Trying again in 5 secs...
Fri Aug  2 18:43:41 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 18:43:41 2024 -> ERROR: Download failed (35) Fri Aug  2 18:43:41 2024 -> ERROR:  Message: SSL connect error
Fri Aug  2 18:43:41 2024 -> ERROR: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 18:43:41 2024 -> Giving up on https://database.clamav.net...
Fri Aug  2 18:43:41 2024 -> daily database available for download (remote version: 27355)
Fri Aug  2 18:43:42 2024 -> WARNING: Download failed (35) Fri Aug  2 18:43:42 2024 -> WARNING:  Message: SSL connect error
Fri Aug  2 18:43:42 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Fri Aug  2 18:43:42 2024 -> Trying again in 5 secs...
```

Edit: confirmed that even after ansible has run the next day, it still hasn't triggered a restart, as the "running since" time is from yesterday.

```
$ systemctl status clamav-freshclam
● clamav-freshclam.service - ClamAV virus database updater
     Loaded: loaded (/lib/systemd/system/clamav-freshclam.service; enabled; ven>
     Active: active (running) since Fri 2024-08-02 17:42:40 UTC; 17h ago
       Docs: man:freshclam(1)
             man:freshclam.conf(5)
             https://docs.clamav.net/
   Main PID: 23549 (freshclam)
      Tasks: 1 (limit: 9506)
     Memory: 5.7M
        CPU: 7.385s
     CGroup: /system.slice/clamav-freshclam.service
             └─23549 /usr/bin/freshclam -d --foreground=true
```

**Comment by jemrobinson** [comment]
@JimMadge : I've got the handler to trigger by setting `changed_when: true` on the task. If there's a better way to indicate that the file has *actually* changed, let me know, otherwise I'll merge as-is.

File: data_safe_haven/resources/workspace/ansible/desired_state.yaml
Comment (JimMadge) [inline data_safe_haven/resources/workspace/ansible/desired_state.yaml]:
```suggestion
```
You shouldn't need that. The module will trigger the handler whenever the file is changed.

Comment (jemrobinson) [reply data_safe_haven/resources/workspace/ansible/desired_state.yaml]: See notes above. The module *doesn't* trigger the handler when the file is changed without this.

Comment (JimMadge) [reply data_safe_haven/resources/workspace/ansible/desired_state.yaml]: I'm about 99.9% certain if the file is updated, the `lineinfile` module will report changes and the handler will be run.

My guess is what happened is the `notify` was added after the file had been edited, so when running the playbook again the module reports OK and the handler isn't called.

Comment (JimMadge) [reply data_safe_haven/resources/workspace/ansible/desired_state.yaml]: Using `changed_when: true` or `changed_when: false` is bad practice because tasks will always report as changed or OK regardless of what the module does.

Comment (jemrobinson) [reply data_safe_haven/resources/workspace/ansible/desired_state.yaml]: As I said above, if you know a better way to trigger the notify if-and-only-if the file is changed, I'm happy to update this.

Comment (JimMadge) [reply data_safe_haven/resources/workspace/ansible/desired_state.yaml]: I might not be explaining very well. I'm sure that is exactly what should happen without any `changed_when` declaration. Like we already do for sshd and auditd.

I'll deploy this branch (without the `changed_when`) and check.

Comment (jemrobinson) [reply data_safe_haven/resources/workspace/ansible/desired_state.yaml]: I agree this *should* happen but it wasn't working for me. The start time of the `clamav-freshclam` service was before the file change was made, and there was no output from the `Restart freshclam` handler in the logs. My theory is that `lineinfile` wasn't triggering the notify - note that neither the sshd and auditd examples use the "copy a line from another file" behaviour, which could be the problem here.

**Comment by jemrobinson** [comment]
Here are some logs from a deployment with `changed_when: true` still enabled. Note that even though the `Restart freshclam` handler has run, it hasn't actually restarted the `freshclam` service. You can see that the service was started at `10:17:56` (just after the `clamav` user was created) and neither the `Enable and start freshclam service` (which should just start it if it isn't started) nor the `Restart freshclam` have restarted it.

```
$ journalctl -u desired-state
Aug 05 10:17:48 shm-pink-sre-fuschia-vm-workspace-03 groupadd[27124]: new group: name=clamav, GID=127
Aug 05 10:17:48 shm-pink-sre-fuschia-vm-workspace-03 useradd[27131]: new user: name=clamav, UID=118, GID=127, home=/var
...
Aug 05 10:21:52 shm-pink-sre-fuschia-vm-workspace-03 desired_state.sh[4903]: TASK [Enable and start freshclam service] ******************
Aug 05 10:21:53 shm-pink-sre-fuschia-vm-workspace-03 desired_state.sh[4903]: TASK [Set freshclam private mirror] ****
...
Aug 05 10:23:25 shm-pink-sre-fuschia-vm-workspace-03 desired_state.sh[4903]: RUNNING HANDLER [Restart freshclam] ***
```

```
$ systemctl status clamav-freshclam
● clamav-freshclam.service - ClamAV virus database updater
     Loaded: loaded (/lib/systemd/system/clamav-freshclam.service; enabled; ven>
     Active: active (running) since Mon 2024-08-05 10:17:56 UTC; 30min ago
       Docs: man:freshclam(1)
             man:freshclam.conf(5)
             https://docs.clamav.net/
   Main PID: 27957 (freshclam)
      Tasks: 1 (limit: 19167)
     Memory: 4.2M
        CPU: 453ms
     CGroup: /system.slice/clamav-freshclam.service
             └─27957 /usr/bin/freshclam -d --foreground=true
```

This is then reflected in the logs, which are still looking at the old download location

```
$ cat /var/log/clamav/freshclam.log 
Mon Aug  5 10:17:56 2024 -> --------------------------------------
Mon Aug  5 10:17:56 2024 -> freshclam daemon 0.103.11 (OS: linux-gnu, ARCH: x86_64, CPU: x86_64)
Mon Aug  5 10:17:56 2024 -> ClamAV update process started at Mon Aug  5 10:17:56 2024
Mon Aug  5 10:17:56 2024 -> daily database available for download (remote version: 27358)
Mon Aug  5 10:17:56 2024 -> WARNING: Download failed (35) Mon Aug  5 10:17:56 2024 -> WARNING:  Message: SSL connect error
Mon Aug  5 10:17:56 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Mon Aug  5 10:17:56 2024 -> Trying again in 5 secs...
Mon Aug  5 10:18:01 2024 -> daily database available for download (remote version: 27358)
Mon Aug  5 10:18:01 2024 -> WARNING: Download failed (35) Mon Aug  5 10:18:01 2024 -> WARNING:  Message: SSL connect error
Mon Aug  5 10:18:01 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Mon Aug  5 10:18:01 2024 -> Trying again in 5 secs...
Mon Aug  5 10:18:06 2024 -> daily database available for download (remote version: 27358)
Mon Aug  5 10:18:06 2024 -> WARNING: Download failed (35) Mon Aug  5 10:18:06 2024 -> WARNING:  Message: SSL connect error
Mon Aug  5 10:18:06 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Mon Aug  5 10:18:06 2024 -> Trying again in 5 secs...
Mon Aug  5 10:18:11 2024 -> daily database available for download (remote version: 27358)
Mon Aug  5 10:18:11 2024 -> WARNING: Download failed (35) Mon Aug  5 10:18:11 2024 -> WARNING:  Message: SSL connect error
Mon Aug  5 10:18:11 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Mon Aug  5 10:18:11 2024 -> Trying again in 5 secs...
Mon Aug  5 10:18:16 2024 -> daily database available for download (remote version: 27358)
Mon Aug  5 10:18:16 2024 -> ERROR: Download failed (35) Mon Aug  5 10:18:16 2024 -> ERROR:  Message: SSL connect error
Mon Aug  5 10:18:16 2024 -> ERROR: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Mon Aug  5 10:18:16 2024 -> Giving up on https://database.clamav.net...
Mon Aug  5 10:18:16 2024 -> daily database available for download (remote version: 27358)
Mon Aug  5 10:18:16 2024 -> WARNING: Download failed (35) Mon Aug  5 10:18:16 2024 -> WARNING:  Message: SSL connect error
Mon Aug  5 10:18:16 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Mon Aug  5 10:18:16 2024 -> Trying again in 5 secs...
Mon Aug  5 10:18:21 2024 -> daily database available for download (remote version: 27358)
Mon Aug  5 10:18:22 2024 -> WARNING: Download failed (35) Mon Aug  5 10:18:22 2024 -> WARNING:  Message: SSL connect error
Mon Aug  5 10:18:22 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Mon Aug  5 10:18:22 2024 -> Trying again in 5 secs...
Mon Aug  5 10:18:27 2024 -> daily database available for download (remote version: 27358)
Mon Aug  5 10:18:27 2024 -> WARNING: Download failed (35) Mon Aug  5 10:18:27 2024 -> WARNING:  Message: SSL connect error
Mon Aug  5 10:18:27 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Mon Aug  5 10:18:27 2024 -> Trying again in 5 secs...
Mon Aug  5 10:18:32 2024 -> daily database available for download (remote version: 27358)
Mon Aug  5 10:18:32 2024 -> WARNING: Download failed (35) Mon Aug  5 10:18:32 2024 -> WARNING:  Message: SSL connect error
Mon Aug  5 10:18:32 2024 -> WARNING: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Mon Aug  5 10:18:32 2024 -> Trying again in 5 secs...
Mon Aug  5 10:18:37 2024 -> daily database available for download (remote version: 27358)
Mon Aug  5 10:18:37 2024 -> ERROR: Download failed (35) Mon Aug  5 10:18:37 2024 -> ERROR:  Message: SSL connect error
Mon Aug  5 10:18:37 2024 -> ERROR: Can't download daily.cvd from https://database.clamav.net/daily.cvd
Mon Aug  5 10:18:37 2024 -> Giving up on https://database.clamav.net...
Mon Aug  5 10:18:37 2024 -> ERROR: Update failed for database: daily
Mon Aug  5 10:18:37 2024 -> ERROR: Database update process failed: Connection failed
Mon Aug  5 10:18:37 2024 -> ERROR: Update failed.
Mon Aug  5 10:18:37 2024 -> --------------------------------------
```

**Comment by JimMadge** [comment]
Snippet from `desired-state.service` logs on fresh deployment.

```
desired_state.sh[4920]: RUNNING HANDLER [Systemd daemon reload] 
desired_state.sh[4920]: RUNNING HANDLER [Restart sshd] 
desired_state.sh[4920]: RUNNING HANDLER [Restart freshclam] 
desired_state.sh[4920]: RUNNING HANDLER [Restart auditd]
```

```
● clamav-freshclam.service - ClamAV virus database updater
     Loaded: loaded (/lib/systemd/system/clamav-freshclam.service; enabled; ven>
     Active: active (running) since Mon 2024-08-05 10:43:44 UTC; 14min ago
```

Restart freshclam was run at 10:49 🤔.

**Comment by jemrobinson** [comment]
See my notes above (it looks like the handler doesn't actually trigger a restart). I'm going to try with an explicit stop of the service before restarting.

**Comment by JimMadge** [comment]
Same thing with the other units
Maybe the module isn't doing what we think it should.

```
    - name: Restart freshclam
      ansible.builtin.systemd:
        name: clamav-freshclam
        enabled: true
        state: restarted
```

Perhaps we should try

```
    - name: Restart freshclam
      ansible.builtin.systemd:
        name: clamav-freshclam
        state: restarted
```

**Comment by jemrobinson** [comment]
Can you try that while I try the other option?

**Comment by JimMadge** [comment]
> Can you try that while I try the other option?

:+1:

> I'm going to try with an explicit stop of the service before restarting.

I think it would be best to avoid that if possible as it would stop the service every day when the service is run and it would mean the play isn't idempotent.

**Comment by JimMadge** [comment]
It works :tada:

```
[ 9572.109954] desired_state.sh[45761]: RUNNING HANDLER [Restart freshclam] ********************************************
[ 9572.362541] desired_state.sh[45761]: changed: [localhost]
```

```
● clamav-freshclam.service - ClamAV virus database updatertatus clamav-freshclam 
     Loaded: loaded (/lib/systemd/system/clamav-freshclam.service; enabled; ven>
     Active: active (running) since Mon 2024-08-05 13:09:16 UTC; 46s ago
```

**Comment by JimMadge** [comment]
@jemrobinson I think this is ready to go now. Are there any other changes you wanted to make?

@craddm Could you review?

**Comment by jemrobinson** [comment]
Cool - I'll do a deploy now

**Review by craddm — APPROVED** [review]
LGTM - assuming James' test works!

**Comment by jemrobinson** [comment]
@JimMadge : Can we wait until I've finished testing before merging?

**Comment by jemrobinson** [comment]
Still not working for me as of 0e63902f8

**Review by jemrobinson — CHANGES_REQUESTED** [review]
Freshclam service isn't being restarted in notify handler.

**Review by jemrobinson — APPROVED** [review]
Working for me :tada:

**Pull request merged**
